src/BridgeFactory.sol

```diff
@@ -36,70 +48,66 @@ contract BridgeFactory is IERC1271, MultiCallContext, TwoStepOwnable {
         _;
     }
 
-    function isValidSignature(bytes32 _hash, bytes calldata _signature)
+    function _verifyRoot(bytes32 root, address pendingOwner_) internal view {
+        bytes32 initHash = _proxyInitHash;
+        address factory = _cachedThis;
+        assembly ("memory-safe") {
+            let ptr := mload(0x40)
+
+            // derive creation salt
+            mstore(0x00, root)
+            mstore(0x14, pendingOwner_)
+            mstore(0x34, chainid())
+            let salt := keccak256(0x00, 0x54)
+
+            // 0xff + factory + salt + hash(initCode)
+            mstore(0x0d, factory)
+            mstore(0x00, 0xff00000000000000)
+            mstore(0x2d, salt)
+            mstore(0x4d, initHash)
+            let computedAddress := keccak256(0x18, 0x55)
+
+            // verify that `salt` was used to deploy `address(this)`
+            if shl(0x60, xor(address(), computedAddress)) {
+                mstore(0x00, 0x1e092104) // selector for `PermissionDenied()`
+                revert(0x1c, 0x04)
+            }
+
+            // restore clobbered memory
+            mstore(0x40, ptr)
+        }
+    }
+
+    function isValidSignature(bytes32 hash, bytes calldata signature)
         external
         view
         override
         onlyProxy
         returns (bytes4)
     {
-        return _isValidSignature(_hash, _signature);
-    }
-
-    function _isValidSignature(bytes32 _leaf, bytes calldata _proof) private view returns (bytes4) {
         bytes32[] calldata proof;
         assembly ("memory-safe") {
-            // _signature is just going to be the proof, then we can read it as so
-            proof.offset := add(0x20, _proof.offset)
+            // signature is just the proof, then we can read it as so
+            proof.offset := add(signature.offset, calldataload(signature.offset))
```

Comment:
Can we not assume here it is going to be `0x20`? 🤔

Reply:
that's true if the abiencoding is strict, but since anybody may call `isValidSignature`, I don't think we can safely assume that